RecoEgamma/PhotonIdentification/plugins/PhotonIDValueMapProducer.cc

```diff
@@ -396,6 +403,11 @@ float PhotonIDValueMapProducer::computeWorstPFChargedIsolation(const reco::Photo
     // Loop over the PFCandidates
     for (auto const& aCCand : chargedCands) {
       auto iCand = aCCand.candidate;
+
+      //Leave out bizarre PF candidates with nan pT
+      if (edm::isNotFinite(iCand->pt()))
```

Comment:
it's not clear to me why this check here is needed if the previous check on line 266 is already introduced.

Reply:
We have 2 different flavors of ChargedHadronIsolation which are used in Photon ID: 
  - [the ordinary one](https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/interface/Photon.h#L523) and 
  - [the one w.r.t worst-vertex](https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/interface/Photon.h#L524).

The first nan-check in L266 is for the ordinary ChargedHadronIso (and also for other PFIso, like neutral hadron iso).

The second nan-check is for ChargedHadronIsolation w.r.t worst vertex. The 2nd check is also needed as we are again looping over all PF cands here.

I can also give an example (in a relval MC event that I have been checking)

Without any nan-check;
```
chargedHadronIso = -nan
chargedHadronWorstVtxIso = -nan
```
With nan-check ONLY in line 266
```
chargedHadronIso = 17.5619
chargedHadronWorstVtxIso = -nan
```
With nan-check in both places
```
chargedHadronIso = 17.5619
chargedHadronWorstVtxIso = 17.5619
```

Reply:
The 2nd nan-check is within the vertex loop; so if we add a log-warning there, then the same warning message for a single nan PFcand will appear for every vertex. That's why I did not include it. It may add too many warning messages for one single issue. 

If there is a nan PFcand in an event, then it's very likely will be reported once already in the first nan-check. I guess that's enough for debugging purposes.. let me know what you think

Reply:
>  it's very likely will be reported once already in the first nan-check

is it very likely or absolutely certain?

Having various isNaN in various places throughout the code can be confusing in the future - a future maintainer may ask if perhaps every function should contain an isNaN check. Perhaps it would be cleaner to prepare a list of PFCandidates that contain no NaNs, report the errors there, and then use that list everywhere downstream in this module coherently?

Reply:
>is it very likely or absolutely certain?

Looking at the code, it seems absolutely certain; because in both places we are looping over the same collection of PF cand, so a nan will be caught in the first check and will raise a warning.

About maintainability of code, I totally see your point, Joosep. We should do better that what I've done in this PR (which is bypassing the issue, and not addressing the root cause of the issue.) Together with PF experts we are working on understanding the real issue (being discussed here: https://github.com/cms-sw/cmssw/issues/39110). 

The reason I'm rushing to fix it for photon is that, if this isn't backported to 12_4_X soon, the big MC production will happen with this bug; ie photons will have nan isolation in some events, which might create issues for physics analysis. 

The issue also happens for data, so a backport to 12_4_X will also fix it for rest of the prompt reco data. 

We can always improve the code in future, it is a continuous process; but to me it seems useful to fix the photon isolation issue asap for reset of the data-taking this year, and upcoming big MC production.

Reply:
Sure, I'm not proposing some major re-engineering, just suggesting a simple change here along the lines of

```
std::vector<reco::PFCandidateRef> pfCandNoNaN;
for (auto pf : pfCandsHandle->ptrs()) {
  if (edm::isNaN(pf->pt())) {
    edm::LogWarning << "NaN at ... , skipping, see issue #xyz ";
  } else {
    pfCandNoNaN.push_back(pf);
  }
}
```
around line 244, before the loop on the PF candidates, and then replace `*pfCandsHandle` with `pfCandNoNaN` elsewhere in the code, so everything is guaranteed to be clean elsewhere with no more checks or warnings needed.

Reply:
ahhh okay, great, thank you, 
I will implement your suggestion and push it soon